src/rfactor/rfactor.py

```diff
@@ -1,134 +1,281 @@
+        - *all_events_cum* (float): Cumulative rain over all events together
+        - *erosivity_cum* (float): Cumulative erosivity over all events together
+
     """
-    filename = lst_inputs[0]
-    path_results = lst_inputs[1]
-
-    check_oct()
-    from oct2py import Oct2Py,Oct2PyError
-    year = filename.stem.split("_")[1]
-    inputdata = np.loadtxt(str(filename.resolve()))
-
-    with Oct2Py() as oc:
-        try:
-            oc.addpath(str(Path(__file__).parent.resolve()))
-            cumEI = oc.core(year, inputdata)
-            filename_out = path_results / (
-                        filename.stem + 'new cumdistr salles.txt')
-            np.savetxt(filename_out, cumEI.T, "%.3f %.2f %.1f")
-            oc.exit()
-        except Oct2PyError as e:
-            raise SystemError(e)
-
-def check_oct():
-    """Check octave installation
+    if ("datetime" not in rain.columns) or ("rain_mm" not in rain.columns):
+        raise RFactorInputException(
+            "DataFrame should contain 'datetime' and 'rain_mm' columns."
+        )
+    # ?TODO -> is the enforcement of a single year required for this implementation
+    if len(rain["datetime"].dt.year.unique()) != 1:  # data of a single year
+        raise RFactorInputException("DataFrame should contain data of a single year.")
+    # ?TODO -> check if we should enforce 10' interval as input? asfreq?
+
+    # mark start of each rain event
+    rain = rain[rain["rain_mm"] > 0.0]  # Only keep measurements with rain
+    rain["event_start"] = False
+    rain.loc[rain["datetime"].diff() >= event_split, "event_start"] = True
+    rain.loc[rain.index[0], "event_start"] = True
+
+    # add an event identifier
+    rain["event_idx"] = rain["event_start"].cumsum()
+
+    # add cumulative rain for each event
+    rain["event_rain_cum"] = rain.groupby("event_idx")["rain_mm"].cumsum()
+
+    # add rain energy for each event
+    rain["event_energy"] = rain.groupby("event_idx")["rain_mm"].transform(
+        rain_energy_per_unit_depth
+    )
+
+    # calculate the maximal rain intensity in 30minutes interval
+    max_intensity_event = (
+        rain.groupby("event_idx")[["datetime", "rain_mm", "event_rain_cum"]]
+        .apply(intensity_method)
+        .rename("max_30min_intensity")
+        .reset_index()
+    )
+    rain = pd.merge(rain, max_intensity_event, how="left")
+
+    # derive event summary
+    columns = ["datetime", "event_rain_cum", "max_30min_intensity", "event_energy"]
+    rain_events = rain.groupby("event_idx")[columns].agg(
+        {
+            "datetime": "first",
+            "event_rain_cum": "last",
+            "max_30min_intensity": "last",
+            "event_energy": "last",
+        }
+    )
+
+    # calculate the erosivity
+    rain_events["erosivity"] = (
+        rain_events["event_energy"] * rain_events["max_30min_intensity"]
+    )
+
+    # cumulative rain over all events
+    # ?TODO - should we do this after or before exclusion of events below threshold?
```

Comment:
I would state before, as we want the shift to be independent from the 'event_threshold' parameters

Reply:
ok